toolchain/driver/compile_subcommand.cpp

```diff
@@ -442,29 +451,102 @@ class CompilationUnit {
   std::unique_ptr<llvm::Module> module_;
 };
 
-CompilationUnit::CompilationUnit(DriverEnv& driver_env,
-                                 const CompileOptions& options,
+// Caches lists that are shared cross-unit.
+class MultiUnitCache {
+ public:
+  explicit MultiUnitCache(
+      const CompileOptions* options,
+      const llvm::ArrayRef<std::unique_ptr<CompilationUnit>> units)
+      : options_(options), units_(units) {}
+
+  auto include_in_dumps() -> llvm::ArrayRef<bool> {
+    CARBON_CHECK(!units_.empty());
+    if (include_in_dumps_.empty()) {
+      BuildIncludeInDumps();
+    }
+    return include_in_dumps_;
+  }
+
+  auto tree_and_subtrees_getters()
+      -> llvm::ArrayRef<Parse::GetTreeAndSubtreesFn> {
+    CARBON_CHECK(!units_.empty());
+    if (tree_and_subtrees_getters_.empty()) {
+      BuildTreeAndSubtreesGetters();
+    }
+    return tree_and_subtrees_getters_;
+  }
+
+ private:
+  auto BuildIncludeInDumps() -> void {
+    CARBON_CHECK(include_in_dumps_.empty());
+    llvm::append_range(include_in_dumps_,
+                       llvm::map_range(units_, [&](const auto& unit) {
+                         return options_->exclude_dump_file_prefix.empty() ||
+                                !unit->input_filename().starts_with(
+                                    options_->exclude_dump_file_prefix);
+                       }));
+  }
+
+  auto BuildTreeAndSubtreesGetters() -> void {
+    CARBON_CHECK(tree_and_subtrees_getters_.empty());
+    llvm::append_range(
+        tree_and_subtrees_getters_,
+        llvm::map_range(units_, [&](const auto& unit) {
+          return unit->has_source() ? unit->get_trees_and_subtrees() : nullptr;
+        }));
+  }
+
+  const CompileOptions* options_;
+
+  // The units being compiled.
+  const llvm::ArrayRef<std::unique_ptr<CompilationUnit>> units_;
+
+  // For each unit, whether it's included in dumps. Used cross-phase.
+  llvm::SmallVector<bool> include_in_dumps_;
+
+  // For each unit, the `TreeAndSubtrees` getter. Used by lowering.
+  llvm::SmallVector<Parse::GetTreeAndSubtreesFn> tree_and_subtrees_getters_;
+};
+
+}  // namespace
+
+CompilationUnit::CompilationUnit(int unit_index, DriverEnv* driver_env,
+                                 const CompileOptions* options,
                                  Diagnostics::Consumer* consumer,
                                  llvm::StringRef input_filename)
-    : driver_env_(&driver_env),
+    : unit_index_(unit_index),
+      driver_env_(driver_env),
       options_(options),
       input_filename_(input_filename),
       vlog_stream_(driver_env_->vlog_stream) {
-  if (vlog_stream_ != nullptr || options_.stream_errors) {
+  if (vlog_stream_ != nullptr || options_->stream_errors) {
     consumer_ = consumer;
   } else {
     sorting_consumer_ = Diagnostics::SortingConsumer(*consumer);
     consumer_ = &*sorting_consumer_;
   }
-  if (options_.dump_mem_usage && IncludeInDumps()) {
+}
+
+auto CompilationUnit::include_in_dumps() -> bool {
```

Comment:
```suggestion
auto CompilationUnit::IncludeInDumps() -> bool {
```

I think, since this is not just an inline getter, and esp since it's in the cpp file

Reply:
Done